subprojects/messaging/src/main/java/org/gradle/internal/serialize/DefaultSerializerRegistry.java

```diff
@@ -19,24 +19,37 @@
 import com.google.common.base.Objects;
 import org.gradle.internal.Cast;
 
+import javax.annotation.concurrent.ThreadSafe;
 import java.util.Arrays;
 import java.util.Comparator;
 import java.util.HashMap;
-import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.LinkedHashSet;
 import java.util.Map;
 import java.util.Set;
-import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListMap;
 
+/**
+ * Default implementation of {@link SerializerRegistry}.
+ *
+ * This class must be thread-safe because multiple tasks may be registering serializable classes concurrently, while other tasks are calling {@link #build(Class)}.
+ */
+@ThreadSafe
 public class DefaultSerializerRegistry implements SerializerRegistry {
-    private final Map<Class<?>, Serializer<?>> serializerMap = new TreeMap<Class<?>, Serializer<?>>(new Comparator<Class<?>>() {
+    private static final Comparator<Class<?>> CLASS_COMPARATOR = new Comparator<Class<?>>() {
         @Override
         public int compare(Class<?> o1, Class<?> o2) {
             return o1.getName().compareTo(o2.getName());
         }
-    });
-    private final Set<Class<?>> javaSerialization = new HashSet<Class<?>>();
+    };
+    private final Map<Class<?>, Serializer<?>> serializerMap = new ConcurrentSkipListMap<Class<?>, Serializer<?>>(CLASS_COMPARATOR);
+
+    // We are using a ConcurrentHashMap here because:
+    //   - We don't want to use a Set with CLASS_COMPARATOR, since that would identity two classes from different classloaders.
```

Comment:
```suggestion
    //   - We don't want to use a Set with CLASS_COMPARATOR, since that would treat two classes with the same name originating from different classloaders as identical, allowing only one in the Set.
```
It took me a moment to understand this, since the word "identity" is typically not used this way.

Couldn't the name of the classloader be added to the comparison in `CLASS_COMPARATOR` to solve this problem and allow the use of a Set here to save the storage used on this map's unused values?  I suppose then there would be more elements added to `serializerMap` if the same class is loaded from different classloaders, which would remove this benefit - I don't know how often to expect that to happen.